[ENH] Implement dynamic clipping to axes limits for 3D plots

scottshambaugh opened the pull request:
Closes https://github.com/matplotlib/matplotlib/issues/25804

## PR summary
This PR introduces "dynamic clipping" of the data in 3D plots, such that data outside the axis limits is not shown. I did a deep dive into stackoverflow and found 12 separate questions asking for this sort of functionality (see that list in [this comment](https://github.com/matplotlib/matplotlib/issues/25804#issuecomment-1723985066)).

All 3D plotting functions now support the `axlim_clip` keyword argument (default `False`), which when `True` will clip the data to the axes view limits, hiding all data outside those bounds. This clipping will be dynamically applied in real time while panning and zooming.

Please note that if one vertex of a line segment or 3D patch is clipped, the entire segment or patch will be hidden. Not being able to show partial lines or patches such that they are "smoothly" cut off at the boundaries of the view box is a limitation of the current renderer.

What's new and new gallery example image (I'm not in love with this example, but it'll do for now):

![3d_clip_to_axis_limits-1](https://github.com/matplotlib/matplotlib/assets/14363975/ccaeb767-9837-4802-ab5e-7975b62ed8e6)

Video of the dynamic behavior, please excuse my sluggish machine:

https://github.com/matplotlib/matplotlib/assets/14363975/3ab251e2-3a1b-4ca1-81a1-34401f0d105e

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

scottshambaugh commented:
TODO: check to see if this fixes https://github.com/matplotlib/matplotlib/issues/8902

QuLogic commented:
> Not being able to show partial lines or patches such that they are "smoothly" cut off at the boundaries of the view box is a limitation of the current renderer.

I don't think that's the case; we can definitely clip partial lines or patches to the Axes. Unless you mean the 3D renderer?

In that case, you would have to do some magic to calculate the clipping path of the Axes3D box with the artist in flattened 2D space, and apply it with `set_clip_path`. I'm not sure if that is worth it or not.

scottshambaugh commented:
Yeah, I mean smooth clipping for the 3D renderer. It'd be a really big effort I think, especially considering (literal) corner cases. This ticket's I think the best to track that ask: https://github.com/matplotlib/matplotlib/issues/8902

scottshambaugh commented:
Tests cleaned up, this is good for review & merge when acceptable

scottshambaugh commented:
Only test failure is this one, unrelated
https://github.com/matplotlib/matplotlib/issues/27635

QuLogic commented:
Should the new argument be keyword-only? It appears to be the case for most of the classes, when it appears after `*`, but not for e.g., `Text3D` or `Line3DCollection`. It also appears before `*args` in the `Axes3D` methods, and it probably should be keyword-only there.

ksunden commented:
```python

import matplotlib.pyplot as plt
import numpy as np

from matplotlib import cm
from mpl_toolkits.mplot3d.axes3d import get_test_data

# set up a figure twice as wide as it is tall
fig = plt.figure(figsize=plt.figaspect(0.5))

# =============
# First subplot
# =============
# set up the axes for the first plot
ax = fig.add_subplot(1, 2, 1, projection='3d')

# plot a 3D surface like in the example mplot3d/surface3d_demo
X = np.arange(-5, 5, 0.25)
Y = np.arange(-5, 5, 0.25)
X, Y = np.meshgrid(X, Y)
R = np.sqrt(X**2 + Y**2)
Z = np.sin(R)
surf = ax.plot_surface(X, Y, Z, rstride=1, cstride=1, cmap=cm.coolwarm,
                       linewidth=0, antialiased=False, axlim_clip=True)
ax.set_zlim(-1.01, 1.01)
fig.colorbar(surf, shrink=0.5, aspect=10)

# ==============
# Second subplot
# ==============
# set up the axes for the second plot
ax = fig.add_subplot(1, 2, 2, projection='3d')

# plot a 3D wireframe like in the example mplot3d/wire3d_demo
X, Y, Z = get_test_data(0.05)
ax.plot_wireframe(X, Y, Z, rstride=10, cstride=10, axlim_clip=True)

plt.show()
```

(Adapted from [this example](https://matplotlib.org/stable/gallery/mplot3d/subplot3d.html#d-plots-as-subplots), just adding the new kwarg)

I think this example is illustrative of the limitations of this approach, in ways that may in fact be able to be better than they are.

In short, the wireframe works really well, but the filled surface exposes some draw ordering problems that we don't see with the argument set to `False`. (The same kinds of problems we see all the time, where something that is "behind" being drawn "above") It is not super clear to me _why_ the nan-ifying/clipping actually causes this, but we don't see them drawn in the "wrong" order without it.


This is relevant because the example in #25804 was using a filled surface, and so I'm not sure this would actually satisfy their usecase.

@scottshambaugh thoughts?

In general I'm willing to accept "This is the best we can do, but yeah its relatively limited utility, but still want to add it", but if you have ideas here, I'd like to hear them.

scottshambaugh commented:
Coming back to this after a while.
@QuLogic I agree, changed these to keyword-only
@ksunden great catch, I fixed the zorder sorting. Your example works for the surfaces now, here's the before and after for just the left plot:
![axlim_clip_before](https://github.com/matplotlib/matplotlib/assets/14363975/1563312c-bd1b-4960-877b-abce55943692)
![axlim_clip_after](https://github.com/matplotlib/matplotlib/assets/14363975/238b5887-f7fe-4509-8990-123694e08b66)

scottshambaugh commented:
Note to self: merge conflict with https://github.com/matplotlib/matplotlib/pull/27385, will have to resolve after the first of these changes goes in.

scottshambaugh commented:
Working through iterating on the test image to resolve minor RMS difference on the surface plot on just the Python 3.12 on macos-14 build. This is rather annoying, as I need to cycle through the CI to test changes I've confirmed that turning off the `axlim_clip` flag does not resolve things. Looks like an instance of https://github.com/matplotlib/matplotlib/issues/1188 potentially? Anyone have advice on working around it?

scottshambaugh commented:
This is looking good and ready for final review.

scottshambaugh commented:
Needs a rebase before merge

scottshambaugh commented:
Rebase complete

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2617,6 +2642,8 @@ def tricontour(self, *args,
             position in a plane normal to *zdir*.
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
+        axlim_clip : bool, default: False
+            Whether to hide lines with a vertex outside the axes view limits.
```
Ditto.

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2671,6 +2699,8 @@ def contourf(self, X, Y, Z, *args, zdir='z', offset=None, **kwargs):
             position in a plane normal to *zdir*.
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
+        axlim_clip : bool, default: False
+            Whether to hide lines with a vertex outside the axes view limits.
```
Ditto

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2709,6 +2739,8 @@ def tricontourf(self, *args, zdir='z', offset=None, **kwargs):
             position in a plane normal to zdir.
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
+        axlim_clip : bool, default: False
+            Whether to hide lines with a vertex outside the axes view limits.
```
Ditto

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2894,6 +2937,8 @@ def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
             When plotting 2D data, the direction to use as z ('x', 'y' or 'z').
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
+        axlim_clip : bool, default: False
+            Whether to hide bars with points outside the axes view limits.
```
Ditto

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2983,6 +3029,9 @@ def bar3d(self, x, y, z, dx, dy, dz, color=None,
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
 
+        axlim_clip : bool, default: False
+            Whether to hide the bars with points outside the axes view limits.
+
```
Ditto

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -3829,6 +3893,9 @@ def stem(self, x, y, z, *, linefmt='C0-', markerfmt='C0o', basefmt='C3-',
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
 
+        axlim_clip : bool, default: False
+            Whether to hide stems that are outside the axes limits.
+
```
Ditto

QuLogic commented on lib/mpl_toolkits/mplot3d/proj3d.py:
```diff
@@ -171,23 +171,42 @@ def _ortho_transformation(zfront, zback):
-    txs = np.ma.masked_array(txs, ~tis)
-    tys = np.ma.masked_array(tys, ~tis)
-    tzs = np.ma.masked_array(tzs, ~tis)
+    tis = np.ma.array(tis)
```
Since `txs`, `tys`, and `tzs` are masked arrays, this should already be a masked array for the `else` case; you only need to change the true side to be `np.ma.ones` instead.

QuLogic commented on lib/mpl_toolkits/mplot3d/proj3d.py:
```diff
@@ -171,23 +171,42 @@ def _ortho_transformation(zfront, zback):
+    # we check each to protect for scalars
+    if np.ma.isMA(vec[0]):
+        txs.mask = vec[0].mask
+        tis.mask = vec[0].mask
```
Why does `tis` only change for the first element?

scottshambaugh replied:
Updated to key off of all elements

QuLogic commented on lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:
```diff
@@ -2138,10 +2177,10 @@ def test_computed_zorder():
         # plot a vector
-        ax.plot((2, 2), (2, 2), (0, 4), c='red', zorder=2)
+        ax.plot((2, 2), (2, 2), (0, 4), c='red', zorder=2, axlim_clip=False)
 
         # plot some points
-        ax.scatter((3, 3), (1, 3), (1, 3), c='red', zorder=10)
+        ax.scatter((3, 3), (1, 3), (1, 3), c='red', zorder=10, axlim_clip=False)
```
Isn't `False` the default?

scottshambaugh replied:
Yup, reverted

scottshambaugh commented on lib/matplotlib/text.py:
```diff
@@ -753,9 +753,16 @@ def draw(self, renderer):
             # position in Text:
-            posx = float(self.convert_xunits(self._x))
-            posy = float(self.convert_yunits(self._y))
+            x, y = self._x, self._y
+            if np.ma.is_masked(x):
+                x = np.nan
+            if np.ma.is_masked(y):
+                y = np.nan
```
I believe this is covered by `ax.text()` in `test_axlim_clip`?

scottshambaugh commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -392,9 +450,13 @@ def do_3d_projection(self):
         """
+        segments = self._segments3d
+        if self._axlim_clip:
+            segments = [np.ma.column_stack([*_viewlim_mask(*zip(*points), self.axes)])
```
Updated to a vectorized version. It's 4 lines now instead of 1, but makes things mostly vectorized.

QuLogic replied:
To be clear, I did mean a vectorized version of `_viewlim_mask`, but I don't remember now if it would be useful in all cases.

scottshambaugh replied:
Ah, I see what you mean. I do think any potential overhead is minimal in vectorizing `_viewlim_mask` per-axis as it is right now, rather than casting everything to a single numpy array.

scottshambaugh commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2573,6 +2595,8 @@ def contour(self, X, Y, Z, *args,
         data : indexable object, optional
             DATA_PARAMETER_PLACEHOLDER
+        axlim_clip : bool, default: False
+            Whether to hide lines with a vertex outside the axes view limits.
```
Updated this and the others

scottshambaugh commented on lib/matplotlib/collections.py:
```diff
@@ -339,7 +339,7 @@ def _prepare_points(self):
             offset_trf = offset_trf.get_affine()
 
-        if isinstance(offsets, np.ma.MaskedArray):
+        if np.ma.isMaskedArray(offsets):
```
Yup, reverted

scottshambaugh commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -427,11 +490,13 @@ def __init__(self, *args, zs=(), zdir='z', **kwargs):
         """
         super().__init__(*args, **kwargs)
-        self.set_3d_properties(zs, zdir)
+        self.set_3d_properties(zs, zdir, axlim_clip)
```
Weird, I think this deserves its own ticket

scottshambaugh replied:
Opened https://github.com/matplotlib/matplotlib/issues/28703

scottshambaugh commented:
Thank you for the review @QuLogic! I believe all your comments are addressed now

scottshambaugh commented:
The docs build error is unrelated:
```
The version switcher "https://output.circle-artifacts.com/output/job/29bcafd2-ffe3-4f64-ba51-f8cfa5bce2b5/artifacts/0/doc/build/html/_static/switcher.json" file cannot be read due to the following error:
HTTPError('404 Client Error: Not Found for url: https://output.circle-artifacts.com/output/job/29bcafd2-ffe3-4f64-ba51-f8cfa5bce2b5/artifacts/0/doc/build/html/_static/switcher.json')
```

scottshambaugh commented:
I think that the docs error isn't actually generic, will investigate.

QuLogic commented:
> /home/circleci/project/lib/mpl_toolkits/mplot3d/art3d.py:docstring of matplotlib.artist.Poly3DCollection.set:45: WARNING: py:meth reference target not found: matplotlib.collections._CollectionWithSizes.set_sizes [ref.meth]

The missing reference handler is now a bit more specific about these things, so if you've added/removed some lines in a docstring, you may have to update the line numbers in the json file to get it to skip again.

scottshambaugh commented:
This is a rather confusing error. It traces back to `./doc/missing-references.json`, which has the below section:

https://github.com/matplotlib/matplotlib/blob/b01462c9ac7b0d3875892d6f40c9213c56f0864f/doc/missing-references.json#L301-L329

... but `set_sizes` is not in the `art3d.py.Poly3DCollection` docstrings so I'm not sure how to adjust the numbering. Any pointers?

I did add 3 lines to the `__init__()` docstring, so will try adding 3 to each of those lines.

QuLogic commented:
It's not complaining about `Poly3DCollection.__init__`, but `Poly3DCollection.set`, which is one of the weird automatic substitutions of all possible keyword arguments, but I'm not sure which one changed. It's probably easiest to build locally with `missing_references_write_json = True` and see what lines changed.

Also, I just noticed this looking at other arguments, but `axlim_clip` should probably get a `.. versionadded::` directive in its docs.

scottshambaugh commented:
@QuLogic I'm not seeing this error on local docs builds. Do you know of another way to track it down?

scottshambaugh commented:
Ok, fixed the issue with `WARNING: py:meth reference target not found`! For future reference, I deleted the two lines, waited for CI to throw the build error, and then re-added the lines with the errored line numbers.

The remaining docs error is unrelated.

@QuLogic I also added in `.. versionadded::` everywhere applicable.

nbfazel commented:
Thank you for addressing this issue!

Which version of matplotlib will incorporate the fix? My current version is 3.9.2.

scottshambaugh commented:
Hi @nbfazel, this will be out in the future 3.10 release.

nbfazel commented:
Thank you!
